rtc-service/servicesmanager/service.go

```diff
@@ -0,0 +1,165 @@
+package servicesmanager
+
+import (
+	"encoding/json"
+	"time"
+
+	"github.com/acmutd/bsg/rtc-service/logging"
+	"github.com/acmutd/bsg/rtc-service/requests"
+	"github.com/acmutd/bsg/rtc-service/response"
+	"github.com/gorilla/websocket"
+)
+
+var (
+	// Length of time to wait for a pong response from the service.
+	PONG_WAIT = 10 * time.Second
+
+	// This is the interval at which the RTC service will send a ping to the service.
+	// To keep the connection alive.
+	//
+	// The value is set to 90% of the PONG_WAIT time.
+	// This is to ensure that the service has enough time to respond to the ping.
+	PING_INTERVAL = (PONG_WAIT * 9) / 10
+)
+
+// List of all services connected to RTC service.
+type ServicesList map[*Service]bool
+
+// Service represents a connected service.
+type Service struct {
+	// Name of the service connected to RTC service.
+	Name string
+
+	// Connection to the service.
+	Connection *websocket.Conn
+
+	// Used to avoid concurrent writes to the websocket connection.
+	Egress chan response.Response
+
+	// Service Manager
+	ServiceManager *ServiceManager
+}
+
+// Creating a new client to communicate with.
+func NewClient(name string, conn *websocket.Conn, manager *ServiceManager) *Service {
+	return &Service{
+		Name:           name,
+		Connection:     conn,
+		Egress:         make(chan response.Response),
+		ServiceManager: manager,
+	}
+}
+
+// Read the incoming messages from the service.
+//
+// This function should be run as a goroutine.
+func (s *Service) ReadMessages() {
+	// Close the connection when the function returns
+	defer func() {
+		s.ServiceManager.RemoveService(s)
+	}()
+
+	// Configure Wait time for Pong response, use Current time + pongWait
+	// This has to be done here to set the first initial timer.
+	if err := s.Connection.SetReadDeadline(time.Now().Add(PONG_WAIT)); err != nil {
+		logging.Error(err)
+		return
+	}
+	// Configure how to handle Pong responses
+	s.Connection.SetPongHandler(s.pongHandler)
+
+	// Loop Forever
+	for {
+		// ReadMessage is used to read the next message in queue
+		// in the connection
+		_, message, err := s.Connection.ReadMessage()
+
+		if err != nil {
+			// If Connection is closed, we will get an error here
+			if websocket.IsUnexpectedCloseError(err, websocket.CloseGoingAway, websocket.CloseAbnormalClosure) {
+				logging.Error("Error reading message: ", err)
+			}
+			break
+		}
+
+		// Unmarshal message into a message struct.
+		var messageStruct requests.Message
+		err = json.Unmarshal(message, &messageStruct)
+		if err != nil {
+			logging.Error("Failed to unmarshal message: ", err)
+			s.Egress <- *response.NewErrorResponse(response.GENERAL, err.Error())
+		} else {
+			// Update the message with the service name.
+			s.Name = messageStruct.ServiceName
```

Comment:
It seems like you did the opposite of what you have for the comment. You update the service name using the message info.